doc/source/rllib/rllib-offline.rst

```diff
@@ -8,217 +8,662 @@ Working With Offline Data
+    algo = config.build()
+    # Load now the PPO-trained `RLModule` to use in recording.
+    algo.restore_from_path(
+        best_checkpoint,
+        # Load only the `RLModule` component here.
+        component=COMPONENT_RL_MODULE,
+    )
 
-The experiences will be saved in compressed JSON batch format:
+    # Run 10 evaluation iterations and record the data.
+    for i in range(10):
+        print(f"Iteration {i + 1}")
+        res_eval = algo.evaluate()
+        print(res_eval)
 
-.. code-block:: text
+    # Stop the algorithm. Note, this is important for when
+    # defining `output_max_rows_per_file`. Otherwise,
+    # remaining episodes in the `EnvRunner`s buffer will not
+    # be written to disk.
+    algo.stop()
 
-    $ ls -l /tmp/cartpole-out
-    total 11636
-    -rw-rw-r-- 1 eric eric 5022257 output-2019-01-01_15-58-57_worker-0_0.json
-    -rw-rw-r-- 1 eric eric 5002416 output-2019-01-01_15-59-22_worker-0_1.json
-    -rw-rw-r-- 1 eric eric 1881666 output-2019-01-01_15-59-47_worker-0_2.json
+.. note:: The stored episode data is formatted as ``binary``. Each episode is converted into its dictionary representation and serialized using ``msgpack-numpy``, 
+    ensuring version compatibility.
 
-Then, we can tell DQN to train using these previously generated experiences with ``"input": "/tmp/cartpole-out"``. We disable exploration since it has no effect on the input:
+RLlib's  recording process is efficient because it utilizes multiple ``OfflineSingleAgentEnvRunner`` instances during evaluation, enabling parallel 
+data writing. You can explore the folder to review the stored Parquet data:
 
-.. code-block:: bash
+.. code-block:: text
 
-    $ rllib train \
-        --run=DQN \
-        --env=CartPole-v1 \
-        --config='{
-            "input": "/tmp/cartpole-out",
-            "explore": false}'
+    $ ls -la /tmp/docs_rllib_offline_recording/cartpole-v1
 
-Off-Policy Estimation (OPE)
----------------------------
+    drwxr-xr-x.  2 user user 540 18. Nov 18:05 run-000001-00006
+    drwxr-xr-x.  2 user user 540 18. Nov 18:05 run-000001-00011
+    drwxr-xr-x.  2 user user 540 18. Nov 18:05 run-000001-00016
+    drwxr-xr-x.  2 user user 540 18. Nov 18:05 run-000001-00021
 
-In practice, when training on offline data, it is usually not straightforward to evaluate the trained policies using a simulator as in online RL. For example, in recommender systems, rolling out a policy trained on offline data in a real-world environment can jeopardize your business if the policy is suboptimal. For these situations we can use `off-policy estimation <https://arxiv.org/abs/1911.06854>`__ methods which avoid the risk of evaluating a possibly sub-optimal policy in a real-world environment.
+.. hint:: RLlib stores records under a folder named by the RL environment. Therein, you see for each ``OfflineSingleAgentEnvRunner`` and write operation 
+    one folder of Parquet files. The write operation count is given in the second numbering. 
```

Comment:
so this is the 6, 11, 16, 21? We should explain what this means and why exactly we get these numbers.

Reply:
I have described in more detailed how these numbers occur. I will also in another PR bring in some changes to the `OfflineSingleAgentEnvRunner`.